File: USACO/Bronze/Sorting&Sets/yearOfTheCow.py
macro = {"Ox": 1, 
        "Tiger": 2, 
        "Rabbit": 3, 
        "Dragon":4, 
        "Snake":5, 
        "Horse":6, 
        "Goat": 7, 
        "Monkey": 8, 
        "Rooster": 9, 
        "Dog": 10, 
        "Pig": 11, 
        "Rat": 12
    }

n = int(input())

soma = 0
cows_years = dict()
cows_diference = dict()
cows_position = dict()

resultado = 0
cows_years["Bessie"] = "Ox"
cows_diference["Bessie"] = 0
cows_position["Bessie"] = 0

for i in range(n):

    entrada = input().split()
    cows_years[entrada[0]] = entrada[4]
    
    if entrada[3] == "previous":
        diferenca = 12 - (macro[entrada[4]] - macro[cows_years[entrada[7]]])

        if entrada[7] != "Bessie":
            if cows_position[entrada[7]] == 1:
                cows_position[entrada[0]] = 1
            else:
                if diferenca > cows_diference[entrada[7]]:
                    cows_position[entrada[0]] = 1
                elif diferenca < cows_diference[entrada[7]]:
                    cows_position[entrada[0]] = -1
                else:
                    cows_position[entrada[0]] = 0
        else:
            cows_position[entrada[0]] = 1

        if cows_position[entrada[7]] == -1:
            cows_diference[entrada[0]] = diferenca - cows_diference[entrada[7]]
        else:
            cows_diference[entrada[0]] = diferenca + cows_diference[entrada[7]]
                
    if entrada[3] == "next":
        diferenca = macro[entrada[4]] - macro[cows_years[entrada[7]]]
        if diferenca < 0:
            diferenca += 12

        if entrada[7] != "Bessie":
            if cows_position[entrada[7]] == -1:
                cows_position[entrada[0]] = -1
            else:
                if diferenca > cows_diference[entrada[7]]:
                    cows_position[entrada[0]] = -1
                elif diferenca < cows_diference[entrada[7]]:
                    cows_position[entrada[0]] = 1
                else:
                    cows_position[entrada[0]] = 0
        else:
            cows_position[entrada[0]] = -1

        if cows_position[entrada[7]] == 1:
            cows_diference[entrada[0]] = cows_diference[entrada[7]] - diferenca
        else:
            cows_diference[entrada[0]] = diferenca + cows_diference[entrada[7]]

print(cows_diference["Elsie"])

File: USACO/Bronze/Sorting&Sets/yearOfTheCow.cpp
#include <bits/stdc++.h>
using namespace std;

const vector<string> ZODIAC{"Ox",   "Tiger",  "Rabbit",  "Dragon", "Snake", "Horse",
                            "Goat", "Monkey", "Rooster", "Dog",    "Pig",   "Rat"};

struct Relation{
    string name;
    int prev;
    int year;
    string relative;
};

int mod(int n, int m){return ((n % m) + m) % m;}

int main() {
    
    int n; cin >> n;

    vector<Relation> relations(n);
    for(Relation &r : relations){
        string unused;
        string prev_str;
        string animal;

        cin >> r.name >> unused >> unused >> prev_str >> animal >> unused >> unused >> r.relative;
        r.prev = prev_str == "previous";
        r.year = find(ZODIAC.begin(), ZODIAC.end(), animal) - ZODIAC.begin();
    }

    map<string, int> birthYears{{"Bessie", 0}};
    for(Relation r : relations){
        int change = r.prev ? -1 : 1;
        int year = birthYears[r.relative] + change;

        while(mod(year, ZODIAC.size()) !=  r.year){year += change;}
        birthYears[r.name] = year;
    }

    int dist = abs(birthYears["Elsie"]);
    cout << dist << endl;
    return 0;
}